homeassistant/components/device_tracker/icloud.py

```diff
@@ -1,100 +1,455 @@
+
+        _CONFIGURING[self.accountname] = configurator.request_config(
+            self.hass, 'iCloud {}'.format(self.accountname),
+            self.icloud_verification_callback,
+            description=('Please enter the validation code:'),
+            description_image="/static/images/config_icloud.png",
+            submit_caption='Confirm',
+            fields=[{'code': '0'}]
+        )
+
+    def keep_alive(self, now):
+        """Keep the api alive."""
+        # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+        from pyicloud.exceptions import PyiCloud2FARequiredError
+
+        if self.api is None:
+            self.reset_account_icloud()
+
+        if self.api is None:
+            return
+
+        if self.api.requires_2fa:
+            try:
+                self.api.authenticate()
+            except PyiCloud2FARequiredError:
+                if self._trusted_device is None:
+                    self.icloud_need_trusted_device()
+                    return
+
+                if self._verification_code is None:
+                    self.icloud_need_verification_code()
+                    return
+
+                if self._verification_code == 'waiting':
+                    return
+
+                if self.api.validate_verification_code(
+                        self._trusted_device, self._verification_code):
+                    self._verification_code = None
+        else:
+            self.api.authenticate()
+
+        currentminutes = dt_util.now().hour * 60 + dt_util.now().minute
+        for devicename in self.devices:
+            interval = self._intervals.get(devicename, 1)
+            if ((currentminutes % interval == 0) or
+                    (interval > 10 and
+                     currentminutes % interval in [2, 4])):
+                self.update_device(devicename)
+
+    def determine_interval(self, devicename, latitude, longitude, battery):
+        """Calculate new interval."""
+        # pylint: disable=too-many-branches
+        distancefromhome = None
+        zone_state = self.hass.states.get('zone.home')
+        zone_state_lat = zone_state.attributes['latitude']
+        zone_state_long = zone_state.attributes['longitude']
+        distancefromhome = distance(latitude, longitude, zone_state_lat,
+                                    zone_state_long)
+        distancefromhome = round(distancefromhome / 1000, 1)
+
+        currentzone = active_zone(self.hass, latitude, longitude)
+
+        if ((currentzone is not None and
+             currentzone == self._overridestates.get(devicename)) or
+                (currentzone is None and
+                 self._overridestates.get(devicename) == 'away')):
+            return
+
+        self._overridestates[devicename] = None
+
+        if currentzone is not None:
+            self._intervals[devicename] = 30
+            return
+
+        if distancefromhome is None:
+            return
+        if distancefromhome > 100:
```

Comment:
Because GMTT doesn't update all the time and could still have used old coordinates from before the update. We would need to be able to call for an update of the gmtt sensor first to be reliable on smaller distances